publish-mqtt/src/main.rs

```diff
@@ -174,90 +173,153 @@ impl Sensor {
+                    .await
+                    .with_context(|| std::line!().to_string())?;
+            }
+
+            {
+                let state = &mut *state.lock().await;
+                connect_first_sensor_in_queue(
+                    bt_session.clone(),
+                    &mut state.homie,
+                    &mut state.sensors_connected,
+                    &mut state.sensors_to_connect,
+                )
+                .await
+                .with_context(|| std::line!().to_string())?;
+            }
+
+            {
+                let state = &mut *state.lock().await;
+                disconnect_first_stale_sensor(
+                    &mut state.homie,
+                    &mut state.sensors_connected,
+                    &mut state.sensors_to_connect,
+                )
+                .await
+                .with_context(|| std::line!().to_string())?;
+            }
+            time::delay_for(CONNECT_INTERVAL).await;
+        }
+        #[allow(unreachable_code)]
+        Ok(())
+    };
+    let t2 = async {
+        service_bluetooth_event_queue(state.clone(), bt_session.clone())
+            .await
+            .with_context(|| std::line!().to_string())?;
+        Ok(())
+    };
+    try_join!(t1, t2).map(|((), ())| ())
+}
+
+async fn check_for_sensors(
+    state: Arc<Mutex<SensorState>>,
+    bt_session: MijiaSession,
+    sensor_names: &HashMap<String, String>,
+) -> Result<(), anyhow::Error> {
+    bt_session.start_discovery().await?;
+
+    let sensors = get_sensors(bt_session.clone())
+        .await
+        .with_context(|| std::line!().to_string())?;
+    let state = &mut *state.lock().await;
+    for props in sensors {
+        // Race Condition: When connecting, we remove from
+        // sensors_to_connect before adding to sensors_connected
+        if sensor_names.contains_key(&props.mac_address)
+            && !state
+                .sensors_to_connect
+                .iter()
+                .chain(state.sensors_connected.iter())
+                .find(|s| s.mac_address == props.mac_address)
+                .is_some()
+        {
+            state
+                .sensors_to_connect
+                .push_back(Sensor::new(props, &sensor_names)?)
+        }
     }
+    Ok(())
 }
 
 async fn connect_first_sensor_in_queue(
-    bt_session: &BluetoothSession,
+    bt_session: MijiaSession,
     homie: &mut HomieDevice,
     sensors_connected: &mut Vec<Sensor>,
     sensors_to_connect: &mut VecDeque<Sensor>,
-) -> Result<(), Box<dyn Error>> {
+) -> Result<(), anyhow::Error> {
     println!("{} sensors in queue to connect.", sensors_to_connect.len());
     // Try to connect to a sensor.
     if let Some(mut sensor) = sensors_to_connect.pop_front() {
         println!("Trying to connect to {}", sensor.name);
-        match connect_start_sensor(bt_session, homie, &mut sensor).await {
+        match connect_start_sensor(bt_session.clone(), homie, &mut sensor).await {
```

Comment:
This is an existing issue so it doesn't need to be fixed in this PR, but there are two different ways connect_start_sensor could fail: there could be an issue with the bluetooth connection, which the way it is handled below is fine for, or there could be an error sending the update to the `HomieDevice` (presumably due to the network connection being dropped), which for consistency should be bubbled up in the same way as disconnect_first_stale_sensor does.

Reply:
Discussed in person, but I think it's worth writing down as well:

I think that this should be covered as part of https://github.com/alsuren/mijia-homie/issues/1 / https://github.com/alsuren/reading-xiaomi-temp/pull/32 (you would set connection_status to `::Connecting { reserved_until: five_minutes_time }` and then keep retrying the connect+subscribe until four_minutes_time and spend up to a minute cleaning up after yourself. The consumer of the connect_and_subscribe() function therefore only needs to know about the Connecting status (and not to try connecting while an attempt is in progress) and shouldn't need to care about any more of the details.